### PR TITLE
[SPARK-46993][SQL] Fix constant folding for session variables

### PR DESCRIPTION
**Pull request by srielau**
### What changes were proposed in this pull request?

Remove the unconditional Alias node generation when resolving a variable reference.

### Why are the changes needed?

An Alias that is not at the top level of an expression blocks constant folding.
Constant folding in turn is a requirement for variables to be usable as an argument to numerous functions, such as from_json().
It also has performance implications.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Existing regression tests in sql-session-variables.sql, added a test to validate the fix.

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ColumnResolutionHelper.scala [hunk 1]
```diff
@@ -312,14 +312,39 @@ trait ColumnResolutionHelper extends Logging with DataTypeErrorsBase {
...
 
-    e.transformWithPruning(_.containsAnyPattern(UNRESOLVED_ATTRIBUTE, TEMP_RESOLVED_COLUMN)) {
-      case u: UnresolvedAttribute =>
-        resolve(u.nameParts).getOrElse(u)
-      // Re-resolves `TempResolvedColumn` as variable references if it has tried to be resolved with
-      // Aggregate but failed.
-      case t: TempResolvedColumn if t.hasTried =>
-        resolve(t.nameParts).getOrElse(t)
+    def innerResolve(e: Expression, isTopLevel: Boolean): Expression = withOrigin(e.origin) {
+      if (e.resolved) return e
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ColumnResolutionHelper.scala:316]: to be consistent with the previous code, we should add one more shortcut:
```
if (!e.containsAnyPattern(UNRESOLVED_ATTRIBUTE, TEMP_RESOLVED_COLUMN)) return e
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ColumnResolutionHelper.scala [hunk 2]
```diff
@@ -312,14 +312,39 @@ trait ColumnResolutionHelper extends Logging with DataTypeErrorsBase {
...
+      val resolved = e match {
+        case u @ UnresolvedAttribute(nameParts) =>
+          val result = withPosition(u) {
+            resolve(nameParts).getOrElse(u) match {
+              // We trim unnecessary alias here. Note that, we cannot trim the alias at top-level,
+              // as we should resolve `UnresolvedAttribute` to a named expression. The caller side
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ColumnResolutionHelper.scala:322]: nit: update the comment to not mention UnresolvedAttribute here

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ColumnResolutionHelper.scala [hunk 3]
```diff
@@ -312,14 +312,39 @@ trait ColumnResolutionHelper extends Logging with DataTypeErrorsBase {
...
+              // can trim the top-level alias if it's safe to do so. Since we will call
+              // CleanupAliases later in Analyzer, trim non top-level unnecessary alias is safe.
+              case Alias(child, _) if !isTopLevel => child
+              case other => other
+            }
+          }
+          result
+
+        // Re-resolves `TempResolvedColumn` if it has tried to be resolved with Aggregate
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ColumnResolutionHelper.scala:331]: nit: let's keep the previous comment

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ColumnResolutionHelper.scala (hunk 2 above)
Comment (srielau) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ColumnResolutionHelper.scala:322]: Why not? It is an UnresolvedAttribute.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
seems the code style check failed

**Comment by cloud-fan** [comment]
thanks, merging to master/3.5!

**Pull request closed without merging**